format/FormatNexusJungfrauExt.py

```diff
@@ -0,0 +1,123 @@
+                for i in range(len(self._slices))
+            ]
+        )
+
+    def _start(self):
+        super(FormatNexusJungfrauExt, self)._start()
+        data_handle = self._reader.entries[0].data[0].handle
+
+        # nexus slices
+        self._slices = self._raw_data._datalists[0]._all_slices
+
+        self._gain = self._slice_correction_array(data_handle["gains"][()])
+        self._pedestal = self._slice_correction_array(data_handle["pedestal"][()])
+        self._pedestalRMS = self._slice_correction_array(data_handle["pedestalRMS"][()])
+
+        self.dxtbx_detector_shape = self._gain[:, 0].shape
+
+    def get_14bit_component(self, index, as_flex=False):
+        """
+        Jungfrau raw data are stored in 16-bit numbers, the 2 significant bits
+        encode the gain mode (a number 0,1,2,3), and the other bits store the measurement
+        (sum of a photon scattering term, and a pedestal term)
+        This function returns the latter term (14-bit measurement)
+        :param index: int, shot index
+        :param as_flex: bool, return as tuple of flex arrays
+        """
+        data_handle = self._reader.entries[0].data[0].handle
+        raw_16bit = data_handle["raw"][index]
+        raw_14bit = pad_stacked_format(raw_16bit & 0x3FFF)
+        raw_14bit = self._slice_correction_array(np.array([raw_14bit]))[:, 0]
+        if as_flex:
+            raw_14bit = tuple([flex.double(d.astype(np.float64)) for d in raw_14bit])
+        return raw_14bit
+
+    def get_pedestal_rms(
+        self, index, as_flex=False, return_gain_modes=False, divide_by_gain=True
+    ):
+        """
+        This method uses the dynamic (per-shot) pixel gain mode to lookup the per-shot pedestal RMS
+        This is important for detailed error analysis of pixel measurements.
+        :param index: image index (0-num_images)
+        :param as_flex: if True, return tuple of flex arrays (like get_raw_data), else return as a numpy array
+        :param return_gain_modes: whether to return the gain modes in addition to the pedestalRMS
+        :param divide_by_gain: normalize the pedestalRMS such that the units are the same as get_raw_data (keV)
+        """
+        if self._pedestalRMS is None:
+            raise AttributeError("There is not pedestalRMS in the master file ... ")
+        data_handle = self._reader.entries[0].data[0].handle
+        raw = data_handle["raw"][index]
+        shot_gain_modes = pad_stacked_format(raw >> 14)
+        shot_gain_modes = self._slice_correction_array(np.array([shot_gain_modes]))
+        shot_gain_modes = shot_gain_modes[:, 0]
+        shot_rms = np.zeros(self.dxtbx_detector_shape, np.float64)
+        for mode in 0, 1, 3:
+            gain_index = mode if mode != 3 else 2  # jungfrau quirk
+            sel = shot_gain_modes == mode  # pixels in the gain mode
+            pedestalRms = self._pedestalRMS[:, gain_index]
+            if divide_by_gain:
+                gain = self._gain[:, gain_index]
```

Comment:
Assume `gain > 0` everywhere?

Reply:
No, gain is negative for two of the gain modes.  In those cases I believe the pedestal is actually bigger than the recorded data values, so the gain corrects the minus sign.  @dermen can correct me if I'm getting this wrong.

Reply:
thats correct! Gain mode 0 is always positive , and the other two are relative gains determined in there calibration procedure.

Reply:
🤷‍♂️

Seems madness to me that the gain i.e. ADU / photon can be a negative value, so really feels like something ... odd happened here. If these were my data I would be asking some questions...

Reply:
I guess gain is a poor term, I would instead call it "the factor that puts the data on a keV scale, after said data has been subtracted by the dark pedestal"